sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw

```diff
@@ -0,0 +1,70 @@
+contract;
+
+use std::{assert::assert, chain::auth::*, constants::ZERO, context::gas, contract_id::ContractId, panic::panic, reentrancy::*, result::*};
+use reentrancy_attacker_abi::Attacker;
+use reentrancy_target_abi::Target;
+
+// Return the sender as a ContractId or panic:
+fn get_msg_sender_id_or_panic(result: Result<Sender, AuthError>) -> ContractId {
+    let mut ret = ~ContractId::from(ZERO);
+    if result.is_err() {
+        panic(0);
+    } else {
+        let unwrapped = result.unwrap();
+        if let Sender::ContractId(v) = unwrapped {
+            ret = v;
+        } else {
+            panic(0);
+        };
+    };
+
+    ret
```

Comment:
I know these are tests so style doesn't matter too much, but if we want to encourage a Rusty way of using Sway then I'd try to avoid mutable variables, and `unwrap()` as much as possible.

```suggestion
    match result {
        Result::Ok(Sender::ContractId(v)) => v,
        _ => panic(0),
    }
```

This is completely untested, I've edited this in a GitHub text box, so don't use it.  I have a horrible feeling that it wouldn't type check either?  Ideally `panic()` returns T which matches whatever it has to so it will type check, knowing that it actually never returns, but this sounds like too advanced behaviour for Sway.

To make it type check you could use `{ panic(0); ~ContractId::from(ZERO) }` for the failure arm, though that's annoying.

Reply:
I don't think that's possible yet, hence why `if let` was implemented first as an easy way to unblock `Result.unwrap()`.

Reply:
What's not possible?  That match could also be an `if let`:
```
    if let Result::Ok(Sender::ContractId(v)) = result {
        v
    } else {
        panic(0);
        ~ContractId::from(ZERO)
    }
```
Is that pattern not supported?

Reply:
IIRC that `match` version you wrote isn't possible yet. The `if let` version is, and is definitely a cleanup!

Reply:
I would _love_ to be using match statements, and I think style always matters. 
I have a feeling I'll have to use the workaround you suggested but I'm ok with that, it's still way cleaner.

As for the unwrap(), in this case it seemed ok, as the result is a `Result<Enum<T, T>, Error>`.
The first call to `is_err()` should handle the `Err` case ( in this case with a panic which is fine), and then I'm unwrapping the Sender enum so I can extract the inner value. 

But overall I'll use `match` over something like this anyday.

Reply:
wouldn't the implicit return of the ContractId in 
```
} else {
        panic(0);
        ~ContractId::from(ZERO)
    }
```

be dead code?

Reply:
Yep, but I don't think it will type-check without it.

Reply:
https://github.com/FuelLabs/sway/pull/964

Reply:
Oh yeah.  So panic() should type check after all.  Cool.
```
    if let Result::Ok(Sender::ContractId(v)) = result {
        v
    } else {
        panic(0);
    }
```
But only if that `if let` pattern match will work.

Reply:
fixed.